### PR TITLE
ERCOT Forecast and Actual Capacity

### PR DESCRIPTION
**Pull request by WillKoehrsen**
- Ercot Actual and Forecasted Capacity
- Data from https://www.ercot.com/api/1/services/read/dashboards/supply-demand.json shown on https://www.ercot.com/gridmktinfo/dashboards/supplyanddemand

### CODE REVIEW

File: gridstatus/ercot.py [hunk 1]
```diff
@@ -671,6 +682,115 @@ def _handle_load_forecast(self, doc, forecast_type, verbose=False):
 
         return df
 
+    def get_committed_capacity(self, date="latest", verbose=False):
```
Comment (WillKoehrsen) [inline gridstatus/ercot.py:685]: Example output:

```python
from gridstatus.ercot import Ercot

Ercot.get_committed_capacity()

| Interval Start            | Interval End              | Capacity | 
|---------------------------|---------------------------|----------| 
| 2024-03-27 00:00:00-05:00 | 2024-03-27 00:05:00-05:00 | 52931    | 
| 2024-03-27 00:05:00-05:00 | 2024-03-27 00:10:00-05:00 | 53126    | 
| 2024-03-27 00:10:00-05:00 | 2024-03-27 00:15:00-05:00 | 53393    | 
| 2024-03-27 00:15:00-05:00 | 2024-03-27 00:20:00-05:00 | 53471    | 
| 2024-03-27 00:20:00-05:00 | 2024-03-27 00:25:00-05:00 | 53472    | 
| 2024-03-27 00:25:00-05:00 | 2024-03-27 00:30:00-05:00 | 53610    | 
| 2024-03-27 00:30:00-05:00 | 2024-03-27 00:35:00-05:00 | 53744    | 
| 2024-03-27 00:35:00-05:00 | 2024-03-27 00:40:00-05:00 | 53909    | 
| 2024-03-27 00:40:00-05:00 | 2024-03-27 00:45:00-05:00 | 54048    | 
| 2024-03-27 00:45:00-05:00 | 2024-03-27 00:50:00-05:00 | 54066    | 
```

File: gridstatus/ercot.py [hunk 2]
```diff
@@ -671,6 +682,115 @@ def _handle_load_forecast(self, doc, forecast_type, verbose=False):
...
+            data.loc[
+                # Actual values
+                data["forecast"] == 0,
+                ["Interval Start", "Interval End", "capacity"],
+            ]
+            .rename(columns={"capacity": "Capacity"})
+            .reset_index(drop=True)
+        )
+
+    def get_forecasted_capacity(self, date="latest", verbose=False):
```
Comment (WillKoehrsen) [inline gridstatus/ercot.py:704]: Example output:

```python
from gridstatus.ercot import Ercot

Ercot().get_forecasted_capacity()

| Interval Start            | Interval End              | Publish Time              | Committed Capacity | Available Capacity | 
|---------------------------|---------------------------|---------------------------|--------------------|--------------------| 
| 2024-03-27 15:15:00-05:00 | 2024-03-27 15:20:00-05:00 | 2024-03-27 15:10:02-05:00 | 63770              | 64067.0            | 
| 2024-03-27 15:20:00-05:00 | 2024-03-27 15:25:00-05:00 | 2024-03-27 15:10:02-05:00 | 63849              | 64443.0            | 
| 2024-03-27 15:25:00-05:00 | 2024-03-27 15:30:00-05:00 | 2024-03-27 15:10:02-05:00 | 63927              | 64818.0            | 
| 2024-03-27 15:30:00-05:00 | 2024-03-27 15:35:00-05:00 | 2024-03-27 15:10:02-05:00 | 64005              | 65193.0            | 
| 2024-03-27 15:35:00-05:00 | 2024-03-27 15:40:00-05:00 | 2024-03-27 15:10:02-05:00 | 64083              | 65568.0            | 
| 2024-03-27 15:40:00-05:00 | 2024-03-27 15:45:00-05:00 | 2024-03-27 15:10:02-05:00 | 64161              | 65943.0            | 
| 2024-03-27 15:45:00-05:00 | 2024-03-27 15:50:00-05:00 | 2024-03-27 15:10:02-05:00 | 64239              | 66318.0            | 
| 2024-03-27 15:50:00-05:00 | 2024-03-27 15:55:00-05:00 | 2024-03-27 15:10:02-05:00 | 64317              | 66693.0            | 
| 2024-03-27 15:55:00-05:00 | 2024-03-27 16:00:00-05:00 | 2024-03-27 15:10:02-05:00 | 64395              | 67068.0            | 
| 2024-03-27 16:00:00-05:00 | 2024-03-27 16:05:00-05:00 | 2024-03-27 15:10:02-05:00 | 64474              | 67444.0            | 
```

File: gridstatus/ercot.py [hunk 3]
```diff
@@ -671,6 +682,115 @@ def _handle_load_forecast(self, doc, forecast_type, verbose=False):
...
+                "Interval Start",
+                "Interval End",
+                "Publish Time",
+                "capacity",
+                "forecast",
+                "available",
+            ]
+        ].sort_values("Interval Start")
+
+    def get_available_seasonal_forecasted_capacity(self, date="latest", verbose=False):
```
Comment (WillKoehrsen) [inline gridstatus/ercot.py:758]: Example output:

```python
from gridstatus.ercot import Ercot

Ercot().get_available_seasonal_forecasted_capacity()

| Interval Start            | Interval End              | Publish Time              | Available Capacity | Load Forecast | 
|---------------------------|---------------------------|---------------------------|--------------------|---------------| 
| 2024-03-28 00:00:00-05:00 | 2024-03-28 01:00:00-05:00 | 2024-03-27 15:10:02-05:00 | 60398              | 39601         | 
| 2024-03-28 01:00:00-05:00 | 2024-03-28 02:00:00-05:00 | 2024-03-27 15:10:02-05:00 | 59456              | 38492         | 
| 2024-03-28 02:00:00-05:00 | 2024-03-28 03:00:00-05:00 | 2024-03-27 15:10:02-05:00 | 58245              | 37907         | 
| 2024-03-28 03:00:00-05:00 | 2024-03-28 04:00:00-05:00 | 2024-03-27 15:10:02-05:00 | 57541              | 37650         | 
| 2024-03-28 04:00:00-05:00 | 2024-03-28 05:00:00-05:00 | 2024-03-27 15:10:02-05:00 | 56984              | 38316         | 
| 2024-03-28 05:00:00-05:00 | 2024-03-28 06:00:00-05:00 | 2024-03-27 15:10:02-05:00 | 57026              | 40495         | 
| 2024-03-28 06:00:00-05:00 | 2024-03-28 07:00:00-05:00 | 2024-03-27 15:10:02-05:00 | 57237              | 44125         | 
| 2024-03-28 07:00:00-05:00 | 2024-03-28 08:00:00-05:00 | 2024-03-27 15:10:02-05:00 | 58414              | 46314         | 
| 2024-03-28 08:00:00-05:00 | 2024-03-28 09:00:00-05:00 | 2024-03-27 15:10:02-05:00 | 63320              | 46329         | 
| 2024-03-28 09:00:00-05:00 | 2024-03-28 10:00:00-05:00 | 2024-03-27 15:10:02-05:00 | 70709              | 46095         | 
```

File: gridstatus/ercot.py [hunk 4]
```diff
@@ -533,16 +548,12 @@ def _get_todays_outlook_non_forecast(self, date, verbose=False):
...
             self.default_timezone,
         ), "Only today's data is supported"
-        url = self.BASE + "/supply-demand.json"
 
-        msg = f"Fetching {url}"
-        log(msg, verbose)
+        supply_demand_json = self._get_supply_demand_json(verbose=verbose)
 
-        r = self._get_json(url)
+        date = self._get_date_from_supply_demand_json(supply_demand_json)
```
Comment (kmax12) [inline gridstatus/ercot.py:554]: i dont think date is actually used in this function? not sure why this line was here in the first place

so, we can also get rid of `_get_date_from_supply_demand_json`

File: gridstatus/ercot.py (hunk 2 above)
Comment (kmax12) [inline gridstatus/ercot.py:704]: let's call this `get_capacity_forecast` more consistent with `get_load_forecast`, `get_solar_forecast`, etc

File: gridstatus/ercot.py (hunk 3 above)
Comment (kmax12) [reply gridstatus/ercot.py:758]: same naming comment here `get_available_seasonal_capacity_forecast`

**Review by kmax12 — APPROVED** [review]
LGTM

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -31,7 +31,7 @@ repos:
         additional_dependencies: [".[jupyter]"]
         types_or: [python, jupyter]
   - repo: https://github.com/charliermarsh/ruff-pre-commit
-    rev: v0.0.202
+    rev: v0.3.4
```
Comment (WillKoehrsen) [inline .pre-commit-config.yaml:34]: Upgraded to match the upgraded version in `pyproject.toml`.

File: pyproject.toml [hunk 6]
```diff
@@ -42,7 +42,7 @@ pytest = "^7.1.2"
 pytest-xdist = "^3.0.2"
 pytest-rerunfailures = "^10.3"
 pytest-cov = "^4.0.0"
-ruff = "^0.0.202"
+ruff = "^0.3.4"
```
Comment (WillKoehrsen) [inline pyproject.toml:45]: This needed to be upgraded to work with the [Ruff VSCode Extension](https://marketplace.visualstudio.com/items?itemName=charliermarsh.ruff). AFAICT, the upgrade doesn't affect linting behavior, only how the configuration is specified.

**Pull request merged**